### DISCUSSION TITLE
A simpler way to add logic for filtering with logical operators

### DISCUSSION THREAD

**Discussion by frederikhors** [99designs/gqlgen#2224] [Q&A]
I'm using GraphQL in Golang with [gqlgen](https://github.com/99designs/gqlgen) and [Bun](https://github.com/uptrace/bun) as ORM.

I'm using [DDD](https://en.wikipedia.org/wiki/Domain-driven_design) and [Clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html), meaning that I'm avoiding `one-size-fits-all/lock-in` solutions like Hasura, Postgraphile, DGraph and everything like that; I can (and I eventually will, no hurry) write and organize my code in such a way to achieve their features.

For now I'm wondering about complex queries with filters like: `AND`, `OR`, `NOT` ([better described here](https://github.com/graphile-contrib/postgraphile-plugin-connection-filter/blob/master/docs/examples.md)).

Right now I need to think/write/test code all by myself and this certainly does not speed up production of the software I want to write.

Also, my code will almost certainly be slow, buggy and possibly all wrong as soon as I realize that I will need something extra for this feature.

_After all, this is true for all software and is one of the reasons why open source was born._

After some research I noticed there are two approaches to solving the problem of GraphQL data filtering:

- create all possible combinations for each field of the model ([this is how ent behaves](https://entgo.io/blog/2021/07/01/automatic-graphql-filter-generation) and [other users suggest](https://github.com/99designs/gqlgen/issues/385#issuecomment-432130569))

- create input types that contain the model fields but also logical comparators as a possible value (like [Hasura, Postgraphile and others do](https://hasura.io/docs/latest/graphql/core/databases/postgres/queries/query-filters/))

### QUESTIONS

1. What do you think about? Pros and cons of each choice?

1. If we aren't using Ent as ORM (which already manages this with code generation) can we have a "global" way to map our Graphql/Gqlgen-generated structs to our domain code without manually write that logic by ourself?

1. Has anyone faced this problem before? @RichardLindhout?

**Comment by RichardLindhout** [comment]
Sort of (in gqlgen-sqlboiler), really like what we have now, but complex queries are not possible with the way we do it.
<img width="784" alt="Schermafbeelding 2022-05-31 om 20 26 40" src="https://user-images.githubusercontent.com/6492229/171258647-4a935364-48ac-4cc5-8ad8-a6fb3e8a0c6e.png">

we can do subqueries though on relationships
<img width="527" alt="Schermafbeelding 2022-05-31 om 20 27 47" src="https://user-images.githubusercontent.com/6492229/171258810-e8e9caa9-a352-47b7-bca9-27bb31b57d52.png">

**Comment by RichardLindhout** [comment]
It's hard to structure and/or combinations I don't how to structure that. e.g. subqueries and things like that. Also keep in mind that you do the authorization part of the query as most important part and the user generated filter query in a AND (..generated-filter here)

**Comment by RichardLindhout** [comment]
For complex or/and queries like below I don't really like the appraoch whe have now but for simple things it's ok ;)

<img width="649" alt="Schermafbeelding 2022-05-31 om 20 31 40" src="https://user-images.githubusercontent.com/6492229/171259459-6316a2ae-0d3d-4e89-bc02-3ba5dcd8ce5b.png">
o
